[SPARK-6030][CORE] Using simulated field layout method to compute class shellSize

advancedxy opened the pull request:
SizeEstimator gives wrong result for Integer on 64bit JVM with UseCompressedOops on, this pr fixes that. For more details, please refer [SPARK-6030](https://issues.apache.org/jira/browse/SPARK-6030)
@sryza, I noticed there is a pr to expose SizeEstimator, maybe that should be waited by this pr get merged if we confirm this problem.
And @shivaram would you mind to review this pr since you contribute related code. Also cc to @srowen and @mateiz

advancedxy committed:
Don't alignSize for objects' shellSize, alignSize when added to state.size. Add
some primitive wrapper objects size tests.

advancedxy commented on core/src/test/scala/org/apache/spark/util/SizeEstimatorSuite.scala:
```diff
@@ -59,6 +59,13 @@ class SizeEstimatorSuite
     assertResult(48)(SizeEstimator.estimate(new DummyClass4(new DummyClass3)))
   }
+  
+  test("primitive wrapper objects") {
```
Should I just use primitive wrapper classes or create dummy classes?

AmplabJenkins commented:
Can one of the admins verify this patch?

shivaram commented:
Jenkins, this is ok to test

SparkQA commented:
[Test build #27992 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27992/consoleFull) for   PR 4783 at commit [`8a127dc`](https://github.com/apache/spark/commit/8a127dc55ab1e8d88e940384e3b37839b45fda97).
 * This patch merges cleanly.

shivaram commented:
@advancedxy -- Thanks for the detailed writeup and the gist. I can see the problem but I am not sure this is the complete fix.  You are right that shellSizes don't need to be aligned to 8 bytes for superclasses and the alignment only needs to happen on object boundaries.  However I am not sure this completely extends all the way -- i.e. when the super class has say a 1 byte field then I don't think we can start counting from 13 (12 for Object + 1 for byte).

While looking around I found this very nice tool called the Java Object Layout [1] which prints out the layout with alignment, gaps given an object name. You can see an example of the alignment gaps I am talking about in [2]  where super-classes with boolean fields end up being aligned to 4 bytes by hotspot (I have no idea what the IBM / Oracle JVMs do).

I'll try to think if there is a more elegant fix and if not we can go with this.

cc @rxin

[1] http://openjdk.java.net/projects/code-tools/jol/
[2] http://hg.openjdk.java.net/code-tools/jol/file/6df1a9b2e5da/jol-samples/src/main/java/org/openjdk/jol/samples/JOLSample_06_Gaps.java

advancedxy commented:
Ok, Thanks for the update. I will look into the jol and the example when I get some spare time.

SparkQA commented:
[Test build #27992 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27992/consoleFull) for   PR 4783 at commit [`8a127dc`](https://github.com/apache/spark/commit/8a127dc55ab1e8d88e940384e3b37839b45fda97).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27992/
Test FAILed.

advancedxy commented:
@shivaram -- I followed the [JDK-8024912](https://bugs.openjdk.java.net/browse/JDK-8024912), [JDK-8024913](https://bugs.openjdk.java.net/browse/JDK-8024913) and [JDK-7007564](https://bugs.openjdk.java.net/browse/JDK-7007564), It looks like non-static field size is aligned to wordSize(8 bytes on 64bit JVM without UseCompressedOops, 4bytes otherwise, verified by the Sizeof code I mentioned on the jira). And the hotspot JVM wouldn't change that in short time.

This is the hotspot version, and I have no ideas about other version's of JVM. But I think we can use this setting up for now just like we did with the ALIGN_SIZE? 

And this is what I get for now. If we (including others or some JVM experts) can agree on that, I will update the code. So, waiting for others' opinion now

shivaram commented:
@advancedxy -- Yeah it sounds good to me to use 4-byte shell size alignment without CompressedOops and 8 bytes otherwise (we can wait to see if there are other opinions). If we want to be extra careful we can put this inside a `if(isHotspot)` check.

BTW it would be great if you can check if this is consistent across Java 1.6, 1.7, 1.8 HotSpot versions.

advancedxy commented:
@shivaram I updated the gist, you can look at the result. [gist](https://gist.github.com/advancedxy/2ae7c9cc7629f3aeb679)

Also you can just download the shell script and try it yourself. The script assumes you are on a Mac and installed java 6,7,8. And of course, you should look at first.

``` shell
curl -o align-test.sh https://gist.githubusercontent.com/advancedxy/2ae7c9cc7629f3aeb679/raw/4f23d55d0f836c2d49f05378398f3c01d7ecdab2/align-test.sh && sh align-test.sh
```

advancedxy committed:
primitiveSize(cls) can just return Int. Use a simplified class field layout
method to calculate class instance size. Will add more documents and test cases.
Add a new alignSizeUp function which uses bitwise operators to speedup.

SparkQA commented:
[Test build #28259 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28259/consoleFull) for   PR 4783 at commit [`4b8aa90`](https://github.com/apache/spark/commit/4b8aa90da94d1b416f2d7144222c102f9cef1b3b).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28259 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28259/consoleFull) for   PR 4783 at commit [`4b8aa90`](https://github.com/apache/spark/commit/4b8aa90da94d1b416f2d7144222c102f9cef1b3b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28259/
Test FAILed.

advancedxy committed:
Add some detailed comments in the code. Add some test cases. It's very
difficult to design test cases as the final object alignment will hide a lot of
filed layout details if we just considering the whole size.

SparkQA commented:
[Test build #28277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28277/consoleFull) for   PR 4783 at commit [`a55f3e1`](https://github.com/apache/spark/commit/a55f3e1655b3b6ebef1f0d1bc615654a6803f7c8).
 * This patch merges cleanly.

advancedxy commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -257,21 +262,56 @@ private[spark] object SizeEstimator extends Logging {
+    // The real world field layout is much more complicated. There are three kinds of fields
+    // order in Java 8. And we don't consider the @contended annotation introduced by Java 8.
+    // see the HotSpot classloader code, layout_fields method for more details.
+    // hg.openjdk.java.net/jdk8/jdk8/hotspot/file/tip/src/share/vm/classfile/classFileParser.cpp
```
It exceeds the 100 characters. So, I remove the first "http://" characters.

advancedxy commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -283,4 +323,7 @@ private[spark] object SizeEstimator extends Logging {
     if (rem == 0) size else (size + ALIGN_SIZE - rem)
   }
+
+  private def alignSizeUp(size: Long, alignSize: Int): Long =
```
I think this code is faster than the previous one and it's self-documentary. Do I need to replace the original one?

shivaram replied:
Yeah its fine to replace the original alignSize with this

advancedxy commented:
@shivaram, I add a bunch of comments in the code. I think it's time for you to review it now. And the previous failures are not related to this code change. Wonder why that would happen.
Also, I think you should review the wording in the comments as I am not a native speaker.

SparkQA commented:
[Test build #28277 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28277/consoleFull) for   PR 4783 at commit [`a55f3e1`](https://github.com/apache/spark/commit/a55f3e1655b3b6ebef1f0d1bc615654a6803f7c8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28277/
Test FAILed.

shivaram commented:
Jenkins, retest this please

SparkQA commented:
[Test build #28351 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28351/consoleFull) for   PR 4783 at commit [`a55f3e1`](https://github.com/apache/spark/commit/a55f3e1655b3b6ebef1f0d1bc615654a6803f7c8).
 * This patch merges cleanly.

shivaram commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -49,6 +49,11 @@ private[spark] object SizeEstimator extends Logging {
   private val DOUBLE_SIZE  = 8
 
+  // Fields can be primitive types, sizes are: 1, 2, 4, 8. Or fields can be pointers. The size of 
+  // a point is 4 or 8 depends on the JVM (32-bit or 64-bit) and UseCompressedOops flag.
```
Nits:

size of point -> size of a pointer
depends on the JVM -> depending on the JVM

shivaram commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -257,21 +262,56 @@ private[spark] object SizeEstimator extends Logging {
     var shellSize = parent.shellSize
     var pointerFields = parent.pointerFields
+    
+    // we have only four sizes: 1, 2, 4, 8. pointerSize can be 4 or 8.
```
I don't think you need this comment. Its pretty self explanatory that we are counting how many times each fieldSize + pointers occur

shivaram commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -257,21 +262,56 @@ private[spark] object SizeEstimator extends Logging {
     }
 
-    shellSize = alignSize(shellSize)
+    // Based on the simulated field layout code in Aleksey Shipilev's report:
```
You can delete the two lines 
```
Aleksey Shiplev also leaders the development of JOL (Java Object 
Layout) tool. The simulated code is also in the JOL. 
```

shivaram commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -257,21 +262,56 @@ private[spark] object SizeEstimator extends Logging {
+    // The code is in Figure 9. Aleksey Shiplev also leaders the development of JOL (Java Object 
+    // Layout) tool. The simulated code is also in the JOL. The simplified idea of field layout 
+    // consists of these parts (see more details in the report):
+    // 1. field alignment: HotSpot lay out the fields aligned by their size.
```
reword to - `HotSpot lays out fields aligned by their size`

shivaram commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -257,21 +262,56 @@ private[spark] object SizeEstimator extends Logging {
+    // 1. field alignment: HotSpot lay out the fields aligned by their size.
+    // 2. object alignment: HotSpot rounds instance size up to 8 bytes
+    // 3. consistent field layouts throughout the hierarchy: This means we should layout 
+    // superclass first. And we use superclass's shellSize as a start point to layout the
```
`as a starting point`

shivaram commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -257,21 +262,56 @@ private[spark] object SizeEstimator extends Logging {
+    var alignedSize = shellSize
+    for (size <- fieldSizes if sizeCount(size) > 0) {
+      val count = sizeCount(size)
+      // If there are internal gaps, smaller filed can fit in.
```
filed -> field

shivaram commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -257,21 +262,56 @@ private[spark] object SizeEstimator extends Logging {
+    for (size <- fieldSizes if sizeCount(size) > 0) {
+      val count = sizeCount(size)
+      // If there are internal gaps, smaller filed can fit in.
+      alignedSize = (alignSizeUp(shellSize, size) + size * count) max alignedSize
```
This is a code style comment -- In Spark we typically avoid infix operators (https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-InfixMethods)

You can make this `max(alignedSize, ...)`

SparkQA commented:
[Test build #28351 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28351/consoleFull) for   PR 4783 at commit [`a55f3e1`](https://github.com/apache/spark/commit/a55f3e1655b3b6ebef1f0d1bc615654a6803f7c8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28351/
Test FAILed.

advancedxy committed:
Fix typos and update wording in comments. Using alignSizeUp to compute
alignSize.

SparkQA commented:
[Test build #28363 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28363/consoleFull) for   PR 4783 at commit [`cb0af95`](https://github.com/apache/spark/commit/cb0af956daba1ccd9d2633ccdde20163def1ec68).
 * This patch merges cleanly.

advancedxy commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -49,6 +49,11 @@ private[spark] object SizeEstimator extends Logging {
+
   // Alignment boundary for objects
   // TODO: Is this arch dependent ?
   private val ALIGN_SIZE = 8
```
It's almost always 8 bytes. But, it can be set to other bytes by -XX:ObjectAlignmentInBytes=  . See http://hg.openjdk.java.net/jdk8/jdk8/hotspot/file/87ee5ee27509/test/gc/TestObjectAlignment.java, But I'd rather to deal with that in another pr. I am feeling it's a bit of more work here.

advancedxy commented:
@shivaram I update the comment, please review again. And still unrelated test failure.

SparkQA commented:
[Test build #28363 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28363/consoleFull) for   PR 4783 at commit [`cb0af95`](https://github.com/apache/spark/commit/cb0af956daba1ccd9d2633ccdde20163def1ec68).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28363/
Test FAILed.

advancedxy commented:
ping @shivaram, @srowen and @mateiz

srowen commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -279,8 +315,8 @@ private[spark] object SizeEstimator extends Logging {
+  private def alignSize(size: Long): Long = alignSizeUp(size, ALIGN_SIZE)
+
+  private def alignSizeUp(size: Long, alignSize: Int): Long =
+    (size + alignSize - 1) & ~(alignSize - 1)
```
Can you perhaps explain this computation? I get what it does, but it's not obvious.

advancedxy replied:
I noticed this code when I looked at the HotSpot JVM classloader code. It's a c macro. And I think it's quite elegant. So I ported this computation. I will add some comment if needed.

srowen commented on core/src/main/scala/org/apache/spark/util/SizeEstimator.scala:
```diff
@@ -257,21 +262,52 @@ private[spark] object SizeEstimator extends Logging {
+    shellSize = alignedSize
+
+    // round up instance filed blocks
+    shellSize = alignSizeUp(shellSize, pointerSize)
```
Can the first arg just be `alignedSize`? I found it slightly funny to see the variable assigned twice.
Otherwise looking good to me though I'm not so familiar with this change.

advancedxy replied:
Yes. I thought just using the alignedSize in the first arg before. But like the comment said, It has two steps. The first assignment is to get the internal aligned size. The second assignment is to round up instance field blocks. I thought it wold be more clear to assigned twice. I will change it if you think it's funny.

advancedxy committed:
Merge two shellSize assignments into one. Add some explanation to alignSizeUp
method.

SparkQA commented:
[Test build #28486 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28486/consoleFull) for   PR 4783 at commit [`9f92469`](https://github.com/apache/spark/commit/9f924696b3ad536f474d24657302e01e96eacf47).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28486 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28486/consoleFull) for   PR 4783 at commit [`9f92469`](https://github.com/apache/spark/commit/9f924696b3ad536f474d24657302e01e96eacf47).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28486/
Test FAILed.

shivaram commented:
@advancedxy Do the tests pass on your local machine ? It looks the ExternalSorter uses a size-tracking hash map which in turn depends on the SizeEstimator. So I am wondering if the test failure is actually related to this PR.

advancedxy commented:
I just run test-only for SizeEstimatorSuite. It do pass the tests. I thought the ExternalSorter was unrelated. I will test the ExternalSorter.

advancedxy commented:
@shivaram ExternalSorterSuite failed on my local machine too. Do you have any thought why ExternalSortedSuite is failing.

shivaram commented:
I don't know much about the ExternalSorter. You could try to walk through the code and try to see how the size estimator changes are affecting it.

@mateiz @rxin might know how it depends on the SizeEstimator

advancedxy commented:
@shivaram The reason why ExternalSorter failed is that It doesn't spilling files for these two failure tests. 
( If we increase the input from `0 until 100000` to `0 until 200000`, It will spilling files to disks, which passes the tests) However the input type for sorter is Iterator[(Int, Int)], and the older SizeEstimator gives 32 bytes and the new SizeEstimator gives the same 32 bytes for (Int, Int) (64 bit JVM with UseCompressedOops on assumed). So, It's very wried to see different results.
Seems @mateiz is busy. @jerryshao cloud you take a look at the failed tests, since you wrote some of the tests.

And, I believe there is another bug in the current SizeEstimator, Scala specialized Int, Long, Float, Double for Tuples, So, The size of (Int, Int) should be 24 bytes, rather than 32bytes. Verified by the method introduced by this article http://www.javaworld.com/article/2077496/testing-debugging/java-tip-130--do-you-know-your-data-size-.html.

I will take a look at the size problem.

advancedxy commented:
@shivaram @srowen Tuple2(Int, Int) got specialized to Tuple2$mcII$sp class. But the Tuple2$mcII$sp is a subclass of Tuple2. So in our implementation, the specialized class will get two additional object references. (_1, _2 in superclass Tuple2, in our case). So, for Tuple2(Int, Int), SizeEstimator will give 32 bytes rather than 24 bytes. In theory, the Tuple2(1,2) class filed layout should be something like below.
```
scala.Tuple2$mcII$sp object internals:
OFFSET SIZE TYPE DESCRIPTION VALUE
0 4 (object header) 01 00 00 00 (0000 0001 0000 0000 0000 0000 0000 0000)
4 4 (object header) 00 00 00 00 (0000 0000 0000 0000 0000 0000 0000 0000)
8 4 (object header) 05 c3 00 f8 (0000 0101 1100 0011 0000 0000 1111 1000)
12 4 Object Tuple2._1 null
16 4 Object Tuple2._2 null
20 4 int Tuple2$mcII$sp._1$mcI$sp 1
24 4 int Tuple2$mcII$sp._2$mcI$sp 2
28 4 (loss due to the next object alignment)
Instance size: 32 bytes (reported by Instrumentation API)
Space losses: 0 bytes internal + 4 bytes external = 4 bytes total
```

But in practice, the size of Tuple2(1, 2) is 24 bytes. So is there any scala expert we can ping? I really want to know why Tuple2(1, 2) can be 24 bytes when the specialized version is involved.

rxin commented:
What do you mean in practice Tuple2 is 24 bytes? How did you measure that? If the object layout tool shows it takes 32 bytes, doesn't it just take 32 bytes?

advancedxy commented:
@rxin as I mentioned in the previous comment, I use the method in this article.
http://www.javaworld.com/article/2077496/testing-debugging/java-tip-130--do-you-know-your-data-size-.html

The related code is listed below. The scala library should be included in the classpath
```
// ----------------------------------------------------------------------------
/**
 * A simple class to experiment with your JVM's garbage collector
 * and memory sizes for various data types.
 *
 * @author <a href="mailto:vlad@trilogy.com">Vladimir Roubtsov</a>
 */
import scala.Tuple2;
import scala.Tuple2$mcII$sp;
public class Sizeof
{
    public static class DummyClass1 {}
    public static class DummyClass2 extends DummyClass1 {
        public boolean x;
    }
    public static class DummyClass3 extends DummyClass2 {
        public boolean y;
    }

    public static void main (String [] args) throws Exception
    {
        // "warm up" all classes/methods that we are going to use:
        runGC ();
        usedMemory ();

        // array to keep strong references to allocated objects:
        final int count = 10000; // 10000 or so is enough for small ojects
        Object [] objects = new Object [count];


        long heap1 = 0;

        // allocate count+1 objects, discard the first one:
        for (int i = -1; i < count; ++ i)
        {
            Object object;

            // INSTANTIATE YOUR DATA HERE AND ASSIGN IT TO 'object':

            object = new Tuple2(1, 2); // This gives 24 bytes. 64 bit vm
            // object = new Tuple2$mcII$sp(1, 2); // This gives 32 bytes. 64 bit vm

            // This gives 56 bytes per object.
            // 24B(Tuple2) + 16B(Integer)+ 16B(Integer) = 56B
            // object = new Tuple2(2 * i + 1, 2 * i + 2);

            if (i >= 0)
                objects [i] = object;
            else
            {
                object = null; // discard the "warmup" object
                runGC ();
                heap1 = usedMemory (); // take a "before" heap snapshot
            }
        }

        runGC ();
        long heap2 = usedMemory (); // take an "after" heap snapshot:

        final int size = Math.round (((float)(heap2 - heap1))/count);
        System.out.println ("'before' heap: " + heap1 +
                            ", 'after' heap: " + heap2);
        System.out.println ("heap delta: " + (heap2 - heap1) +
            ", {" + objects [0].getClass () + "} size = " + size + " bytes");
    }

    // a helper method for creating Strings of desired length
    // and avoiding getting tricked by String interning:
    public static String createString (final int length)
    {
        final char [] result = new char [length];
        for (int i = 0; i < length; ++ i) result [i] = (char) i;

        return new String (result);
    }

    // this is our way of requesting garbage collection to be run:
    // [how aggressive it is depends on the JVM to a large degree, but
    // it is almost always better than a single Runtime.gc() call]
    private static void runGC () throws Exception
    {
        // for whatever reason it helps to call Runtime.gc()
        // using several method calls:
        for (int r = 0; r < 4; ++ r) _runGC ();
    }

    private static void _runGC () throws Exception
    {
        long usedMem1 = usedMemory (), usedMem2 = Long.MAX_VALUE;

        for (int i = 0; (usedMem1 < usedMem2) && (i < 1000); ++ i)
        {
            s_runtime.runFinalization ();
            s_runtime.gc ();
            Thread.currentThread ().yield ();

            usedMem2 = usedMem1;
            usedMem1 = usedMemory ();
        }
    }

    private static long usedMemory ()
    {
        return s_runtime.totalMemory () - s_runtime.freeMemory ();
    }

    private static final Runtime s_runtime = Runtime.getRuntime ();

} // end of class
// ----------------------------------------------------------------------------
```

I took another look at the result. I think I got fooled by the result of Tuple2(1,2);  It gives 24 bytes as result, but the javac don't pick the specialized version class. The Tuple2$mcII$sp(1, 2) does give 32 bytes as result. So, The (1, 2) in Scala on 64 bit JVM do takes 32 bytes size. We get that right

The last thing left is to figure out why the ExternalSort keeping failing tests.

advancedxy committed:
The size of Integer class changes from 24 to 16 on a 64-bit JVM with
-UseCompressedOops flag on after the fix. I don't how 100000 was originally
calculated, It looks like 100000 is the magic number which makes sure spilling.
Because of the size change, It fails because there is no spilling at all. Change
the number to a slightly larger number fixes that.

advancedxy commented:
Ping @shivaram, @rxin. I finally got some time to look at the code.
The ExternalSorterSuite kept failing because there was no spilling when we just insert 100000 elements. 
I don't know how 100000 was calculated, maybe you guys can give some related info. Before this pr, The size of  an integer instance is 24 bytes on a 64-bit JVM with UseCompressedOops on. Now, It's a correct 16 bytes. The ExternalSorterSuit inserts integers into buffers, so a reduced size of integer class results a smaller total estimated size, hence no spilling. 

Another Note: When I was digging the ExternalSorterSuite and ExternalSorter, I noticed something wrong. The ExternalSorter uses the SizeTrackingPairBuffer, SizeTrackingPairBuffer[(Int, Int), Int] to be specifically in the test suite. The SizeTrackingPairBuffer[(Int, Int), Int] increases only 40 bytes per insertion(estimated by SizeEstimator). The 40 bytes  consisted of 24 bytes of Tuple2 and 16 bytes Integer. I changes the insertion number based on this fact (24 + 24 per insertion to 24 + 16 per insertion, 48 * 100000 = 40 * 120000). However 24 bytes of Tuple2 is not right, It should be the specialized version of Tuple2[Int, Int] which is 32 bytes. It's not SizeEstimator's fault as It can only get the Tuple2 class rather than scala.Tuple2$mcII$sp. I doubt it may be something related to the scala compiler, I will look into this problem or someone more familiar with scala compiler should.

As for this pr, I think it's good to go. Hope we can get it merged before 1.4

shivaram commented:
Jenkins, test this please

AmplabJenkins commented:
Can one of the admins verify this patch?

shivaram commented:
Jenkins, ok to test

SparkQA commented:
[Test build #31061 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31061/consoleFull) for   PR 4783 at commit [`db1e948`](https://github.com/apache/spark/commit/db1e948097b202573cac16a23a8bf22f1d6e2a5b).

SparkQA commented:
[Test build #31061 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31061/consoleFull) for   PR 4783 at commit [`db1e948`](https://github.com/apache/spark/commit/db1e948097b202573cac16a23a8bf22f1d6e2a5b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

shivaram commented:
Thanks @advancedxy for fixing the tests. This change LGTM. @rxin @srowen Could you also take a final look at this ?

advancedxy commented:
ping @rxin or @srowen .

srowen commented on core/src/test/scala/org/apache/spark/util/SizeEstimatorSuite.scala:
```diff
@@ -59,6 +68,22 @@ class SizeEstimatorSuite
+    assertResult(16)(SizeEstimator.estimate(new java.lang.Short("1")))
+    assertResult(16)(SizeEstimator.estimate(new java.lang.Integer(1)))
+    assertResult(24)(SizeEstimator.estimate(new java.lang.Long(1)))
+    assertResult(16)(SizeEstimator.estimate(new java.lang.Float(1.0)))
```
Don't change this, but FYI 1.0f is a float and 1.0 is a double.

advancedxy replied:
Thanks.  I do know 1.0 is a double. Maybe I didn't think clearly when I was writing the test cases. Just leave it?

srowen commented on core/src/test/scala/org/apache/spark/util/SizeEstimatorSuite.scala:
```diff
@@ -135,5 +160,20 @@ class SizeEstimatorSuite
     assertResult(64)(SizeEstimator.estimate(DummyString("a")))
     assertResult(64)(SizeEstimator.estimate(DummyString("ab")))
     assertResult(72)(SizeEstimator.estimate(DummyString("abcdefgh")))
+    
```
I'm worried that setting the system props above like `os.arch` is going to pollute the JVM state for other tests. It should be restored.

advancedxy replied:
I didn't consider that. But after a look at the code, It seems the `ResetSystemProperties` trait already takes care of that.

srowen commented:
Needs a rebase.

I know it's not changed in this PR, but why are all those fields `var`? they can be initialized once if the `initialize()` method is removed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31548 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31548/consoleFull) for   PR 4783 at commit [`3f80640`](https://github.com/apache/spark/commit/3f80640932c159ef3a92c83527a83f400f5ebc51).

advancedxy commented:
`initialize()` is invoked multiple times in the test suite. Maybe it's a hook for tests....

srowen commented:
OK that all looks good. Let's see what Jenkins says.

advancedxy commented:
A closer look at the `ResetSystemProperties` and `SizeEstimatorSuite`, the `beforeEach` in `ResetSystemProperties` is overriden by `SizeEstimatorSuite`. Should call `super.beforeEach()` in `SizeEatimatorSuite` to be stackable. 

@JoshRosen, you added the `ResetSystemProperties` trait, should we add the ` super.beforeEach()` in the `beforeEach()` method in the `SizeEstimatorSuite`?

SparkQA commented:
[Test build #31548 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31548/consoleFull) for   PR 4783 at commit [`3f80640`](https://github.com/apache/spark/commit/3f80640932c159ef3a92c83527a83f400f5ebc51).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class ExecutorUIData(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31548/
Test PASSed.

JoshRosen commented:
@advancedxy I can't answer the `ResetSystemPrroperties` question off the top of my head and am a bit busy with other stuff right now, so why not try writing a simple standalone example with some printlns or something to figure out the proper override order, etc?

advancedxy commented:
@JoshRosen @srowen, I already tested in the test suite. Just want to make sure if I don't miss anything obvious.
Here is the standalone reproduce sample.  https://gist.github.com/advancedxy/08a76fb618a3f804e108
If that's the case, I believe there is the BlockManagerSuite to be changed too. @JoshRosen when would you have time to look at that? I can send a pr for that if you are busy.

advancedxy committed:
Add super.beforeEach in the beforeEach method to make the trait stackable..
Remove useless leading whitespace.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31624 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31624/consoleFull) for   PR 4783 at commit [`c4dcb41`](https://github.com/apache/spark/commit/c4dcb411ce38eff25204dff487335e39b986cb05).

SparkQA commented:
[Test build #31624 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31624/consoleFull) for   PR 4783 at commit [`c4dcb41`](https://github.com/apache/spark/commit/c4dcb411ce38eff25204dff487335e39b986cb05).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31624/
Test FAILed.

srowen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31663 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31663/consoleFull) for   PR 4783 at commit [`c4dcb41`](https://github.com/apache/spark/commit/c4dcb411ce38eff25204dff487335e39b986cb05).

SparkQA commented:
[Test build #31663 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31663/consoleFull) for   PR 4783 at commit [`c4dcb41`](https://github.com/apache/spark/commit/c4dcb411ce38eff25204dff487335e39b986cb05).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31663/
Test PASSed.

srowen commented:
LGTM
